src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Services/ExtensionAssemblyManager.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Reflection;
+using System.Runtime.Loader;
+using Microsoft.Extensions.Logging;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Services;
+
+internal sealed class ExtensionAssemblyManager
+{
+    private readonly ImmutableDictionary<string, AssemblyLoadContext> _directoryLoadContexts;
+
+    public ExtensionAssemblyManager(ImmutableDictionary<string, AssemblyLoadContext> directoryLoadContexts)
+    {
+        _directoryLoadContexts = directoryLoadContexts;
+    }
+
+    public static ExtensionAssemblyManager Create(ImmutableArray<string> extensionAssemblyPaths, string? devKitDependencyPath, ILoggerFactory loggerFactory)
+    {
+        var logger = loggerFactory.CreateLogger<ExportProviderBuilder>();
+        if (devKitDependencyPath != null)
+        {
+            // Ensure the Roslyn DevKit assemblies are loaded into the default ALC so that extensions can use them.
+            LoadDevKitAssemblies(devKitDependencyPath, logger);
```

Comment:
I wonder if we can just hook the AssemblyLoadContext resolve event and look for these as needed rather than paying the load cost right away?

Reply:
Maybe... I can check if I can hook into the default ALC resolve.